crates/kitsune_p2p/proxy/src/wire.rs

```diff
@@ -0,0 +1,302 @@
+
+/// Type used to correlate proxy message requests / responses.
+#[derive(
+    Debug,
+    Clone,
+    Copy,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Hash,
+    Deref,
+    AsRef,
+    From,
+    Into,
+    serde::Serialize,
+    serde::Deserialize,
+)]
+pub struct MsgId(pub u64);
+
+static NEXT_MSG_ID: std::sync::atomic::AtomicU64 = std::sync::atomic::AtomicU64::new(0);
+
+impl MsgId {
+    /// Generate the next process-unique channel id.
+    pub fn next() -> Self {
+        Self(NEXT_MSG_ID.fetch_add(1, std::sync::atomic::Ordering::Relaxed))
+    }
+}
+
+/// Type used to denote a logical proxy channel.
+#[derive(
+    Debug,
+    Clone,
+    Copy,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Hash,
+    Deref,
+    AsRef,
+    From,
+    Into,
+    serde::Serialize,
+    serde::Deserialize,
+)]
+pub struct ChannelId(pub u64);
+
+static NEXT_CHANNEL_ID: std::sync::atomic::AtomicU64 = std::sync::atomic::AtomicU64::new(0);
+
+impl ChannelId {
+    /// Generate the next process-unique channel id.
+    pub fn next() -> Self {
+        Self(NEXT_CHANNEL_ID.fetch_add(1, std::sync::atomic::Ordering::Relaxed))
+    }
+}
+
+/// Type used for content data of wire proxy messages.
+#[derive(Debug, PartialEq, Deref, AsRef, From, Into, serde::Serialize, serde::Deserialize)]
+pub struct ChannelData(#[serde(with = "serde_bytes")] pub Vec<u8>);
+
+/// Wire type for transfering urls.
+#[derive(Debug, Clone, PartialEq, PartialOrd, Hash, serde::Serialize, serde::Deserialize)]
+pub struct WireUrl(String);
+
+impl WireUrl {
+    /// Convert to url2.
+    pub fn to_url2(&self) -> url2::Url2 {
+        self.into()
+    }
+
+    /// Convert to url2.
+    pub fn into_url2(self) -> url2::Url2 {
+        self.into()
+    }
+}
+
+macro_rules! q_from {
+    ($($t1:ty => $t2:ty, | $i:ident | {$e:expr},)*) => {$(
+        impl From<$t1> for $t2 {
+            fn from($i: $t1) -> Self {
+                $e
+            }
+        }
+    )*};
+}
+
+q_from! {
+         String => WireUrl,      |s| { Self(s) },
+        &String => WireUrl,      |s| { Self(s.to_string()) },
+           &str => WireUrl,      |s| { Self(s.to_string()) },
+     url2::Url2 => WireUrl,    |url| { Self(url.to_string()) },
+    &url2::Url2 => WireUrl,    |url| { Self(url.to_string()) },
+        WireUrl => url2::Url2, |url| { url2::url2!("{}", &url.0) },
+       &WireUrl => url2::Url2, |url| { url2::url2!("{}", &url.0) },
+}
+
+#[cfg(test)]
+macro_rules! test_val {
+    ($($t:ty {$v:expr},)*) => {
+        trait TestVal {
+            fn test_val() -> Self;
+        }
+
+        $(
+            impl TestVal for $t {
+                fn test_val() -> Self {
+                    $v
+                }
+            }
+        )*
+    };
+}
+
+macro_rules! write_proxy_wire {
+    ($(
+        $(#[doc = $doc:expr])* $s_name:ident :: $c_name:ident($b:literal) {$(
+            $(#[doc = $t_doc:expr])* ($t_name:ident :: $t_idx:tt): $t_ty:ty,
+        )*},
+    )*) => {
```

Comment:
Took a while to understand, but cool :+1:

Reply:
added:

```rust
/// This macro DRYs out implementing the wire protocol variants
/// as there is a lot of shared code between them.
///
/// DSL:
///
/// $s_name - snake-case name
/// $c_name - camel-case name
/// $b      - protocol variant identifier byte (u8) literal
/// $t_name - type name (snake-case)
/// $t_idx  - type index in the message array
/// $t_ty   - type rust type
///
/// Docs allowed on variant and types.
///
/// E.g.:
///
/// write_proxy_wire! {
///    /// Forward data through the proxy channel.
///    /// Send zero length data for keep-alive.
///    chan_send::ChanSend(0x30) {
///        /// The channel id to send data through.
///        (channel_id::0): ChannelId,
///
///        /// The data content to be sent.
///        (channel_data::1): ChannelData,
///    },
/// }
```